gensim/models/doc2vec.py

```diff
@@ -794,6 +794,11 @@ def train(self, documents=None, corpus_file=None, total_examples=None, total_wor
 
         """
         kwargs = {}
+
+        # Check the type of corpus_file
+        if not isinstance(corpus_file, string_types):
```

Comment:
`corpus_file` may legitimately be None, when `documents` is not None. This is the reason why some of the unit tests fail. Please have a look at Travis CI.

So, if you go ahead with your proposed check, you need something like:

```suggestion
        if corpus_file is not None and not isinstance(corpus_file, string_types):
```

Also, please add a unit test that stresses your new functionality (pass in a non-string `corpus_file`, expect a `TypeError` raised).

Next, what is the benefit of raising TypeError here? What happens with the existing code if we do _not_ raise TypeError? What kind of exception does the user see?

I think if we go ahead with this kind of parameter checking, we should do it properly:

1. Ensure that one of `documents` or `corpus_file` is None (they cannot both be non-None)
2. Ensure that one of `documents` or `corpus_file` is not None (they cannot both be None)
3. If `documents` is not None, then it must be an iterable
4. If `corpus_file` is not None, then it must be a string

My question is: is it worth it? @piskvorky

Reply:
Agreed on the tests. I edited the PR description for context (link to original issue).

Reply:
@piskvorky @mpenkov Thanks for the detailed feedback. Working on it.

Reply:
Also I'd say: while checking whether it's a string may help a bit, if there's checking, it'd make sense to ensure things like illegal or missing paths also generate meaningful error messages. It might be possible to just leverage some existing path-checking/file-existence-checking method, or wrap the failing method(s) in a handler that catches errors and shows a good-enough error message like "Problem with `corpus_file` value X".

Reply:
@gojomo so should I check `corpus_file` for being a valid file path instead of checking for `string_type` ?

Reply:
I'm not sure, what covers the most cases with the most straightforward code? A path-test, with the right catches/error-message, might "catch N birds with one stone", pointing the user in the right direction to understand their parameter error with fewer conditionals/lines-of-tests. Or it might complicate things compared to the simple proper-type test. But the motivating case for this change - that users following slightly-older examples would get a confusing error message, due to a new `corpus_file` parameter` – will actually already have been handled by the simple "one or the other but not both" test. So, even the string test isn't strictly required to address the motivating case. My pref would be: do the simplest thing that resolves the motivating case, for sure. If there's a easy/clear bit of extra checking that makes also makes sense, consider it as well.